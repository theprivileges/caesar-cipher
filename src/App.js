/* @flow */
import * as React from 'react';
import makeStyles from '@material-ui/core/styles/makeStyles';
import Container from '@material-ui/core/Container';
import Grid from '@material-ui/core/Grid';
import Typography from '@material-ui/core/Typography';
import Paper from '@material-ui/core/Paper';
import { encrypt } from './utils/index'
import Shift from './Shift';
import PlainText from './PlainText';
import EncryptedText from './EncryptedText';

const { useReducer } = React;

type ThemeObject = {
    spacing: (number) => string,
    palette: {
        text: {
            secondary: string
        }
    }
};

type useStylesObject = {
    paper: {
        padding: string,
        color: string,
    },
    delta: {
        padding: string,
        textAlign: string,
    }
};

type useStylesFunction = (void) => { delta: string, paper: string};

const makeStylesCb = (theme: ThemeObject) : useStylesObject => ({
    delta: {
        padding: theme.spacing(2),
        textAlign: 'center',
    },
    paper: {
        padding: theme.spacing(2),
        color: theme.palette.text.secondary,
    },
})


const useStyles: useStylesFunction = makeStyles(makeStylesCb);

const initialState = {
    shift: 0,
    text: '',
    encrypted: '',
};

// Actions
const SHIFT_UPDATE = 'SHIFT_UPDATE';
const TEXT_UPDATE = 'TEXT_UPDATE';
const ENCRYPTED_UPDATE = 'ENCRYPTED_UPDATE';

// Reducer
const reducer = (state, action) => {
    switch(action.type) {
        case SHIFT_UPDATE:
            const { shift } = action.payload;
            return { shift, text: '', encrypted: ''};
        case TEXT_UPDATE:
            const { text } = action.payload;
            return { ...state, text };
        case ENCRYPTED_UPDATE:
            const { encrypted } = action.payload;
            return { ...state, encrypted };
        default:
            return {...state};
    }
};

// Action Creators
const updateShift = (shift) => ({
    type: SHIFT_UPDATE,
    payload: { shift },
});

const updateText = (text) => ({
    type: TEXT_UPDATE,
    payload: { text },
});

const updateEncrypted = (encrypted) => ({
    type: ENCRYPTED_UPDATE,
    payload: { encrypted }
});


function App(): React.Node {
    const classes = useStyles();
    const [state, dispatch] = useReducer(reducer, initialState);
    const { text, encrypted, shift } = state;

    const handleDeltaChange = (e: SyntheticInputEvent<HTMLSelectElement>) : void => {
        const value = Number.parseInt(e.target.value, 10);
        dispatch(updateShift(value))
    };

    const handleTextChange = (e: SyntheticInputEvent<HTMLInputElement>) : void => {
        const { value } = e.target;
        dispatch(updateText(value))
        dispatch(updateEncrypted(encrypt(value, shift)));
    };

    const handleEncryptedChange = (e: SyntheticInputEvent<HTMLInputElement>) : void => {
        const { value } = e.target;
        const unshift = (26 - shift) % 26;
        dispatch(updateEncrypted(value));
        dispatch(updateText(encrypt(value, unshift)));
    };

    return (
        <Container size="sm">
            <Typography align="center" component="h1" variant="h2" gutterBottom>
                Caesar Cipher
            </Typography>
            <Grid container spacing={3}>
                <Grid item xs={12}>
                    <Paper className={classes.delta} elevation={0}>
                        <Shift value={shift} onChange={handleDeltaChange}/>
                    </Paper>
                </Grid>
                <Grid item xs={12} sm={6}>
                    <Paper className={classes.paper} elevation={0}>
                        <PlainText value={text} onChange={handleTextChange}/> 
                    </Paper>
                </Grid>
                <Grid item xs={12} sm={6}>
                    <Paper className={classes.paper} elevation={0}>
                        <EncryptedText value={encrypted} onChange={handleEncryptedChange}/>
                    </Paper>
                </Grid>
            </Grid>
        </Container>
  );
}

export default App;
