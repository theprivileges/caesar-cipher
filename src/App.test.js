import React from 'react';
import ReactDOM from 'react-dom';
import App from './App';
import renderer from 'react-test-renderer'

describe('App', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
  });
  describe('Snapshots', () => {
    let component;
    beforeEach(() => {
      component = renderer.create(<App />);
    });
    it('renders correctly', () => {
      let tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
})