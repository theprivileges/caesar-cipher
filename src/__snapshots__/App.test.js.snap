// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Snapshots renders correctly 1`] = `
<div
  className="MuiContainer-root MuiContainer-maxWidthLg"
  size="sm"
>
  <h1
    className="MuiTypography-root MuiTypography-h2 MuiTypography-gutterBottom MuiTypography-alignCenter"
  >
    Caesar Cipher
  </h1>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <div
        className="MuiPaper-root makeStyles-delta-260 MuiPaper-elevation0 MuiPaper-rounded"
      >
        <div
          className="MuiFormControl-root"
        >
          <label
            className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
            data-shrink={true}
            htmlFor="delta"
          >
            Delta
          </label>
          <div
            className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
            onClick={[Function]}
          >
            <div
              aria-haspopup="listbox"
              aria-labelledby=" delta"
              className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
              id="delta"
              onBlur={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onMouseDown={[Function]}
              role="button"
              tabIndex={0}
            >
              <em>
                None
              </em>
            </div>
            <input
              name="delta"
              onAnimationStart={[Function]}
              type="hidden"
              value={0}
            />
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root MuiSelect-icon"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
          </div>
          <p
            className="MuiFormHelperText-root MuiFormHelperText-filled"
          >
            How many letters to shift.
          </p>
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6"
    >
      <div
        className="MuiPaper-root makeStyles-paper-261 MuiPaper-elevation0 MuiPaper-rounded"
      >
        <div
          className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
        >
          <label
            className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
            data-shrink={false}
            htmlFor="plain"
            id="plain-label"
          >
            Plain Text
          </label>
          <div
            className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              autoFocus={false}
              className="MuiInputBase-input MuiInput-input"
              disabled={false}
              id="plain"
              name="plain"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6"
    >
      <div
        className="MuiPaper-root makeStyles-paper-261 MuiPaper-elevation0 MuiPaper-rounded"
      >
        <div
          className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
        >
          <label
            className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
            data-shrink={false}
            htmlFor="encrypted"
            id="encrypted-label"
          >
            Encrypted Text
          </label>
          <div
            className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              autoFocus={false}
              className="MuiInputBase-input MuiInput-input"
              disabled={false}
              id="encrypted"
              name="encrypted"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
